File: src/models/session.js
/**
 * 会话管理模块 - 异步 I/O 版本
 */
const fs = require('fs').promises;
const path = require('path');
const crypto = require('crypto');

// 会话数据文件路径
const SESSIONS_FILE = path.join(__dirname, '../../data/sessions.json');

// 会话有效期 (7天)
const SESSION_EXPIRY = 7 * 24 * 60 * 60 * 1000;

// 内存缓存
let sessionsCache = null;
let cacheTime = 0;
const CACHE_TTL = 5000; // 5秒缓存

/**
 * 确保数据目录存在
 */
async function ensureDataDir() {
  const dataDir = path.dirname(SESSIONS_FILE);
  await fs.mkdir(dataDir, { recursive: true }).catch(err => {
    if (err.code !== 'EEXIST') throw err;
  });
}

/**
 * 读取所有会话（带缓存）
 */
async function readSessions() {
  // 缓存命中
  if (sessionsCache !== null && (Date.now() - cacheTime) < CACHE_TTL) {
    return sessionsCache;
  }
  
  try {
    const data = await fs.readFile(SESSIONS_FILE, 'utf8');
    sessionsCache = JSON.parse(data);
    cacheTime = Date.now();
    return sessionsCache;
  } catch (err) {
    if (err.code === 'ENOENT') {
      sessionsCache = {};
      cacheTime = Date.now();
      return sessionsCache;
    }
    throw err;
  }
}

/**
 * 保存所有会话
 */
async function writeSessions(sessions) {
  await ensureDataDir();
  await fs.writeFile(SESSIONS_FILE, JSON.stringify(sessions, null, 2), 'utf8');
  // 更新缓存
  sessionsCache = sessions;
  cacheTime = Date.now();
}

/**
 * 创建会话
 * @param {string} username 用户名
 * @returns {Promise<string>} 会话ID
 */
async function createSession(username) {
  const sessions = await readSessions();
  const sessionId = crypto.randomBytes(32).toString('hex');
  
  sessions[sessionId] = {
    username,
    createdAt: Date.now(),
    expiresAt: Date.now() + SESSION_EXPIRY
  };
  
  await writeSessions(sessions);
  return sessionId;
}

/**
 * 获取会话
 * @param {string} sessionId 会话ID
 * @returns {Promise<object|null>} 会话信息
 */
async function getSession(sessionId) {
  if (!sessionId) return null;
  
  const sessions = await readSessions();
  const session = sessions[sessionId];
  
  if (!session) return null;
  
  // 检查是否过期
  if (Date.now() > session.expiresAt) {
    delete sessions[sessionId];
    await writeSessions(sessions);
    return null;
  }
  
  return session;
}

/**
 * 删除会话
 * @param {string} sessionId 会话ID
 */
async function deleteSession(sessionId) {
  const sessions = await readSessions();
  if (sessions[sessionId]) {
    delete sessions[sessionId];
    await writeSessions(sessions);
  }
}

/**
 * 清理过期会话
 */
async function cleanExpiredSessions() {
  const sessions = await readSessions();
  const now = Date.now();
  let changed = false;
  
  for (const [id, session] of Object.entries(sessions)) {
    if (now > session.expiresAt) {
      delete sessions[id];
      changed = true;
    }
  }
  
  if (changed) {
    await writeSessions(sessions);
    console.log('已清理过期会话');
  }
}

/**
 * 清除缓存
 */
function clearCache() {
  sessionsCache = null;
  cacheTime = 0;
}

// 每小时清理过期会话
setInterval(cleanExpiredSessions, 60 * 60 * 1000);

module.exports = {
  createSession,
  getSession,
  deleteSession,
  cleanExpiredSessions,
  clearCache
};

File: go-impl/internal/model/session.go
// Package model 数据模型
package model

import (
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"sync"
	"sync/atomic"
	"time"
)

// Session 会话信息
type Session struct {
	Username  string `json:"username"`
	CreatedAt int64  `json:"createdAt"`
	ExpiresAt int64  `json:"expiresAt"`
}

// SessionManager 会话管理器
type SessionManager struct {
	filePath    string
	sessions    sync.Map // sessionID -> *Session
	sessionExpy time.Duration
	dirty       atomic.Bool
}

var (
	sessionManager *SessionManager
	sessionOnce    sync.Once
)

// GetSessionManager 获取会话管理器单例
func GetSessionManager() *SessionManager {
	sessionOnce.Do(func() {
		execPath, _ := os.Executable()
		baseDir := filepath.Dir(execPath)
		sessionManager = &SessionManager{
			filePath:    filepath.Join(baseDir, "data", "sessions.json"),
			sessionExpy: 7 * 24 * time.Hour,
		}
		sessionManager.load()
		go sessionManager.startCleanup()
	})
	return sessionManager
}

func (sm *SessionManager) load() {
	data, err := os.ReadFile(sm.filePath)
	if err != nil {
		return
	}

	var sessions map[string]*Session
	if json.Unmarshal(data, &sessions) == nil {
		for id, s := range sessions {
			sm.sessions.Store(id, s)
		}
	}
}

func (sm *SessionManager) save() {
	sessions := make(map[string]*Session)
	sm.sessions.Range(func(key, value interface{}) bool {
		sessions[key.(string)] = value.(*Session)
		return true
	})

	os.MkdirAll(filepath.Dir(sm.filePath), 0755)
	data, _ := json.MarshalIndent(sessions, "", "  ")
	os.WriteFile(sm.filePath, data, 0644)
	sm.dirty.Store(false)
}

// CreateSession 创建会话
func (sm *SessionManager) CreateSession(username string) (string, error) {
	bytes := make([]byte, 32)
	rand.Read(bytes)
	sessionID := hex.EncodeToString(bytes)

	now := time.Now()
	sm.sessions.Store(sessionID, &Session{
		Username:  username,
		CreatedAt: now.UnixMilli(),
		ExpiresAt: now.Add(sm.sessionExpy).UnixMilli(),
	})

	sm.dirty.Store(true)
	go sm.save() // 异步保存

	return sessionID, nil
}

// GetSession 获取会话
func (sm *SessionManager) GetSession(sessionID string) *Session {
	if sessionID == "" {
		return nil
	}

	value, ok := sm.sessions.Load(sessionID)
	if !ok {
		return nil
	}

	session := value.(*Session)
	if time.Now().UnixMilli() > session.ExpiresAt {
		return nil
	}

	return session
}

// DeleteSession 删除会话
func (sm *SessionManager) DeleteSession(sessionID string) {
	sm.sessions.Delete(sessionID)
	sm.dirty.Store(true)
	go sm.save()
}

func (sm *SessionManager) startCleanup() {
	ticker := time.NewTicker(time.Hour)
	for range ticker.C {
		sm.cleanup()
	}
}

func (sm *SessionManager) cleanup() {
	now := time.Now().UnixMilli()
	changed := false

	sm.sessions.Range(func(key, value interface{}) bool {
		session := value.(*Session)
		if now > session.ExpiresAt {
			sm.sessions.Delete(key)
			changed = true
		}
		return true
	})

	if changed {
		sm.save()
	}
}
